apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1701,6 +1722,54 @@ private void HandleAckResponse(Task writeTask, List<string> ackIds, List<string>
                 }
                 // Immediately send more data if there is any to send.
                 StartPush();
+                return;
+
+                // This local method gets the status of ACK/NACK response for the successful and permanently failed ids as the sequence of AckNackResponse.
+                // ids - The list of all the ids being ACK-ed/NACK-ed for which the AckNackResponse is required.
+                // writeTask - The task that contains the result of ACK/NACK request for the given ids.
+                // Returns the status of ACK/NACK response for the successful and permanently failed ids as the sequence of AckNackResponse.
+                IReadOnlyList<AckNackResponse> GetAckNackResponses(List<string> ids, Task writeTask)
+                {
+                    // Either the task would have succeeded or faulted. Let's check for faults.
+                    if (writeTask.IsFaulted)
+                    {
+                        // There may be few successes and few failures or all failures.
+                        var rpcException = writeTask.Exception.As<RpcException>();
+                        if (rpcException != null && _exactlyOnceDeliveryEnabled)
+                        {
+                            // Exactly once delivery scenario.
+                            // Populate AckNackResponse with success and permanent failed ids only.
+                            // Temporary errors are deliberately ignored for maintaining consistency with the other client libraries.
+                            // Temporary errors will eventually become successful or fail permanently. Their status would be shared then.
+                            var ackError = AckError.ForRpcException(rpcException, ids);
+                            return ackError.GetAckNackResponses().ToList();
+                        }
+                        else if (rpcException != null)
+                        {
+                            // Non-exactly once delivery scenario with RpcException.
+                            // All RpcExceptions are either recoverable or ignored. Non-RpcException would throw.
+                            // As a requirement from Pub/Sub team, all RpcExceptions in non-exactly once delivery scenario should be treated as fire and forget,
+                            // so we always treat them as successful (even when they are not). If there are recoverable errors, the ACK/NACK operation
+                            // will be retried, so the message may be sent multiple times with success status.
+                            return GetSuccessResponses(ids);
+                        }
+                        else
+                        {
+                            // Non-gRPC error will throw exception eventually, so don't return those IDs as success. Return an empty list.
```

Comment:
Can you explain this?

Reply:
I see, thanks for the clear explanation. Can you give me an example of a non-GRPC error for ack/modack?

Reply:
There isn't one which I have observed or known. Essentially, the code just caters to the case that if non-gRPC error is thrown (if any) what would be the behavior.

Reply:
We go into this path if writeTask is faulted and if `writeTask.Exception.As<RpcException>()` returns null. I believe this should logically not happen since GRPC should return only RpcExceptions. I don't think we can assume that an exception will come up elsewhere.

Since this path should not happen, per my understanding, can we instead log an error when it does? Or we can assert that `writeTask.Exception.As<RpcException>()` always returns non-null.

Reply:
This code change (to only throw non-gRPC errors) was done earlier this year. Please see this [PR and discussion]( https://github.com/googleapis/google-cloud-dotnet/pull/8287#pullrequestreview-927576158). If we need to change this behavior, I am happy to change it. I can club it in the new issue, document the desired expected behavior in the issue and then do it as part of the next PR. Please let me know.

Reply:
Please add a comment with the summary of our offline discussion. We can close this comment after that.

Reply:
As discussed offline, updating the details for non-gRPC errors here. In case of any exception in the response of ack/nack request, we try and cast the exception to RpcException. If that is successful, it is a gRPC error else it will be a non-gRPC error. In case of non-gRPC error, we explicitly throw an exception [Please see here](https://github.com/googleapis/google-cloud-dotnet/pull/8973/files#:~:text=throw%20writeTask.Exception.FlattenIfPossible()%3B). Sharing the status of AckIds/NackIds with non-gRPC error as success would be incorrect as after a while exception would be thrown, hence  rather than passing the Status of these IDs as success, we send an empty list.